pennylane/transforms/metric_tensor.py

```diff
@@ -291,13 +387,247 @@ def processing_fn(probs):
+            ids.extend(_ids)
+
+    # Combine block-diagonal and off block-diagonal tapes
+    tapes = diag_tapes + first_term_tapes
+    # prepare off block-diagonal mask
+    mask = 1 - qml.math.block_diag([qml.math.ones((bsize, bsize)) for bsize in block_sizes])
+    # Required for slicing in processing_fn
+    num_diag_tapes = len(diag_tapes)
+
+    def processing_fn(results):
+        """Postprocessing function for the full metric tensor."""
+        nonlocal mask
+        # Split results
+        diag_res, off_diag_res = results[:num_diag_tapes], results[num_diag_tapes:]
+        # Get full block-diagonal tensor
+        diag_mt = diag_proc_fn(diag_res)
+
+        # Prepare the mask to match the used interface
+        mask = qml.math.convert_like(mask, diag_mt)
+
+        # Initialize off block-diagonal tensor using the stored ids
+        first_term = qml.math.zeros_like(diag_mt)
+        if ids:
+            off_diag_res = qml.math.stack(off_diag_res, 1)[0]
+            inv_ids = [_id[::-1] for _id in ids]
+            first_term = qml.math.scatter_element_add(first_term, list(zip(*ids)), off_diag_res)
+            first_term = qml.math.scatter_element_add(first_term, list(zip(*inv_ids)), off_diag_res)
+
+        # Second terms of off block-diagonal metric tensor
+        expvals = qml.math.zeros_like(first_term[0])
+        idx = 0
+        for prob, obs in zip(diag_res, obs_list):
+            for o in obs:
+                l = qml.math.cast(o.eigvals, dtype=np.float64)
+                w = tape.wires.indices(o.wires)
+                p = qml.math.marginal_prob(prob, w)
+                expvals = qml.math.scatter_element_add(expvals, (idx,), qml.math.dot(l, p))
+                idx += 1
+
+        # Construct <\partial_i\psi|\psi><\psi|\partial_j\psi> and mask it
+        second_term = qml.math.tensordot(expvals, expvals, axes=0) * mask
+
+        # Subtract second term from first term
+        off_diag_mt = first_term - second_term
+
+        # Rescale first and second term
+        _coeffs = qml.math.hstack(coeffs)
+        scale = qml.math.convert_like(qml.math.tensordot(_coeffs, _coeffs, axes=0), results[0])
+        off_diag_mt = scale * off_diag_mt
+
+        # Combine block-diagonal and off block-diagonal
+        mt = off_diag_mt + diag_mt
+
+        return mt
+
+    return tapes, processing_fn
+
+
+def _get_aux_wire(aux_wire, tape):
+    r"""Determine an unused wire to be used as auxiliary wire for Hadamard tests.
+
+    Args:
+        aux_wire (object): Input auxiliary wire. Returned unmodified if not ``None``
+        tape (pennylane.tape.QuantumTape): Tape to infer the wire for
+
+    Returns:
+        object: The auxiliary wire to be used. Equals ``aux_wire`` if it was not ``None`` ,
+        and an often reasonable choice else.
+    """
+    if aux_wire is not None:
+        return aux_wire
+
+    _wires = tape.wires
+    for _aux in range(tape.num_wires):
+        if _aux not in _wires:
+            return _aux

```

Comment:
Wait, this confuses me:

- shouldn't `tape.num_wires == len(tape.wires)`?
- I don't think this would work if the tape or device wires are not consecutive integers?

Reply:
Yes this is a good point 🤔 

The only solution I can think of is:

- `aux` becomes a _compulsory_ argument when calling the `metric_tensor_hadamard_test` function.

- The QNode execution wrapper, which has access to the device, handles the logic for finding a device wire that is not used by the tape. It then calls the transform with this value of `aux`.

That way, the device is used by the QNode wrapper to determine the correct aux wire

Reply:
The adopted solution is to pass `device_wires` optionally to `_get_aux_wire`. The QNode wrapper, which has access to the device, passes `qnode.device.wires` to the tape transform and users may do something similar when manually using the tape transform. If the tape transform is used manually and no device wires are provided, the old behaviour is adopted.

Note the following nice feature: If we have `device_wires` (so at least always when call `metric_tensor` on a QNode), we now also can _check_ whether a user-provided `aux-wire` is in the `device_wires`, leading to a much cleaner interrupt if this is not the case.